contrib/win32/win32compat/misc.c

```diff
@@ -2119,3 +2124,97 @@ strrstr(const char *inStr, const char *pattern)
 
 	return last;
 }
+
+int
+add_mark_of_web(const char* filename)
+{
+	if (motw_zone_id < 5) {
+		char* fileStreamPath = NULL;
+		size_t fileStreamPathLen = strlen(filename) + strlen(":Zone.Identifier") + 1;
+
+		fileStreamPath = malloc(fileStreamPathLen * sizeof(char));
+
+		if (fileStreamPath == NULL) {
+			return -1;
+		}
+
+		sprintf_s(fileStreamPath, fileStreamPathLen, "%s:Zone.Identifier", filename);
+
+		char* zoneIdentifierStr = NULL;
+		size_t zoneIndentifierLen = strlen("[ZoneTransfer]\nZoneId=") + 1 + 1;
+
+		zoneIdentifierStr = malloc(zoneIndentifierLen * sizeof(char));
+
+		if (zoneIdentifierStr == NULL) {
+			return -1;
+		}
+
+		sprintf_s(zoneIdentifierStr, zoneIndentifierLen, "[ZoneTransfer]\nZoneId=%d", motw_zone_id);
+
+		int ofd, status = 0;
+
+		// create zone identifer file stream and then write the Mark of the Web to it
+		if ((ofd = open(fileStreamPath, O_WRONLY | O_CREAT, USHRT_MAX)) == -1) {
+			status = -1;
+			goto cleanup;
+		}
+
+		if (atomicio(vwrite, ofd, zoneIdentifierStr, zoneIndentifierLen) != zoneIndentifierLen) {
+			status = -1;
+		}
+
+		if (close(ofd) == -1) {
+			status = -1;
+		}
+
+	cleanup:
+		free(fileStreamPath);
+		free(zoneIdentifierStr);
```

Comment:
Will might need to check `zoneIdentifierStr` is null before freeing, if we don't do early out above.